File: Neetcode/LinkedListCycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        nodeset = set()
        if head == None:
            return False
        while head:
            if head in nodeset:
                return True
            else:
                nodeset.add(head)
                head = head.next
        return False

File: Neetcode/LinkedListCycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        ListNode* slow = head;
        ListNode* fast = head;
        if (head == nullptr){
            return false;
        }
        while (true){
            slow = slow->next;
            if (slow == nullptr){
                return false;
            }
            fast = fast->next;
            if (fast == nullptr){
                return false;
            }
            fast = fast->next;
            if (fast == nullptr){
                return false;
            }
            if (slow == fast){
                return true;
            }
        }
        return false;
    }
};
